src/common/src/array/map_array.rs

```diff
@@ -0,0 +1,294 @@
+    type ArrayType = MapArray;
+
+    #[cfg(not(test))]
+    fn new(_capacity: usize) -> Self {
+        panic!("please use `MapArrayBuilder::with_type` instead");
+    }
+
+    #[cfg(test)]
+    fn new(capacity: usize) -> Self {
+        Self::with_type(
+            capacity,
+            DataType::Map(MapType::from_kv(DataType::Varchar, DataType::Varchar)),
+        )
+    }
+
+    fn with_type(capacity: usize, ty: DataType) -> Self {
+        let inner = ListArrayBuilder::with_type(capacity, ty.into_map().into_list());
+        Self { inner }
+    }
+
+    fn append_n(&mut self, n: usize, value: Option<MapRef<'_>>) {
+        self.inner.append_n(n, value.map(|v| v.0));
+    }
+
+    fn append_array(&mut self, other: &MapArray) {
+        self.inner.append_array(&other.inner);
+    }
+
+    fn pop(&mut self) -> Option<()> {
+        self.inner.pop()
+    }
+
+    fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    fn finish(self) -> MapArray {
+        let inner = self.inner.finish();
+        MapArray { inner }
+    }
+}
+
+/// `MapArray` is physically just a `List<Struct<key: K, value: V>>` array, but with some additional restrictions
+/// on the `key`.
+#[derive(Debug, Clone, Eq)]
+pub struct MapArray {
+    pub(super) inner: ListArray,
+}
+
+impl EstimateSize for MapArray {
+    fn estimated_heap_size(&self) -> usize {
+        self.inner.estimated_heap_size()
+    }
+}
+
+impl Array for MapArray {
+    type Builder = MapArrayBuilder;
+    type OwnedItem = MapValue;
+    type RefItem<'a> = MapRef<'a>;
+
+    unsafe fn raw_value_at_unchecked(&self, idx: usize) -> Self::RefItem<'_> {
+        let list = self.inner.raw_value_at_unchecked(idx);
+        MapRef(list)
+    }
+
+    fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    fn to_protobuf(&self) -> PbArray {
+        let mut array = self.inner.to_protobuf();
+        array.array_type = PbArrayType::Map as i32;
+        array
+    }
+
+    fn null_bitmap(&self) -> &Bitmap {
+        self.inner.null_bitmap()
+    }
+
+    fn into_null_bitmap(self) -> Bitmap {
+        self.inner.into_null_bitmap()
+    }
+
+    fn set_bitmap(&mut self, bitmap: Bitmap) {
+        self.inner.set_bitmap(bitmap)
+    }
+
+    fn data_type(&self) -> DataType {
+        let list_value_type = self.inner.values().data_type();
+        DataType::Map(MapType::from_list_entries(list_value_type))
+    }
+}
+
+impl MapArray {
+    pub fn from_protobuf(array: &PbArray) -> ArrayResult<ArrayImpl> {
+        let inner = ListArray::from_protobuf(array)?.into_list();
+        Ok(Self { inner }.into())
+    }
+
+    /// Return the inner struct array of the list array.
+    pub fn as_struct(&self) -> &StructArray {
+        self.inner.values().as_struct()
+    }
+
+    /// Returns the offsets of this map.
+    pub fn offsets(&self) -> &[u32] {
+        self.inner.offsets()
+    }
+}
+#[derive(Clone, Eq, EstimateSize)]
+pub struct MapValue(pub(crate) ListValue);
```

Comment:
Should we make the inner value private to better enforce the invariants on it by restricting the places where a new value can be instantiated?

Also, can you add some more documentation discussing all invariants in one place, such as...

- The datatypes and nullability of keys and values.
- The semantics of `Eq`.
- Whether or not the keys are ordered.